Tagged ptrs

gvanrossum opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

gvanrossum committed:
Introducing tagged pointers (no tags yet!)

- _tagged_ptr type (a union of uintptr_t and PyObject*)
- inline functions is_tagged(), detag(), untaggged() tagged(),
  decref_unless_tagged(), incref_if_tagged(), CLEAR_TAGGED()
- accessors localsplus_as_tagged_ptr(f) and localsplus_as_object_ptr(f)
- Change return type of _PyFrame_Stackbase(), _PyFrame_GetStackPointer()
- Change argument type of _PyFrame_SetStackPointer()
- Change implementation of *_PyFrame_StackPeek(), *Pop(), *Push()
- Split _PyFrame_GetLocalsArray() into *_ObjectPtr() and *_TaggedPtr()
- Fix compilation errors in frameobject.c, typeobject.c, frame.c

Note that no values have the tag bit set yet, so this is a no-op.
Also stack_pointer is still a PyObject **, else ceval.c won't compile.

gvanrossum committed:
Make stack_pointer a _tagged_ptr, fix fallout

Most of the fallout is adding tagging/untagging operations to
the code generator.

Since no tags are ever set, we don't know how much of this works.

gvanrossum committed:
Dumb experiment

def f(a):
    return 424242 + a

vs.

def f(a):
    return a + 424242

The latter will crash, the former won't.

gvanrossum committed:
WIP -- 479 errors left

gvanrossum committed:
Use DECREF_INPUTS() more

Also convert END_FOR (the only macro inst) to a regular inst.

gvanrossum committed:
WIP -- 161 errors left -- must fix Py_DECREF(decref_unless_tagged(...))

gvanrossum committed:
WIP -- 138 errors left

gvanrossum committed:
Whoops, cannot use DECREF_INPUTS() in some cases

gvanrossum committed:
WIP -- 96 errors left

gvanrossum committed:
WIP -- 30 errors left

gvanrossum committed:
WIP -- 26 errors left -- change TRACE_FUNCTION_EXIT()

gvanrossum committed:
WIP -- 19 errors left -- use untagged(NULL) to init conditional input

gvanrossum committed:
WIP -- 47 errors left -- an increase due to more correct typing

gvanrossum committed:
Redesign to_text() substitution feature to allow looking ahead

gvanrossum committed:
Don't substitute detag(x) in LHS of assignment

gvanrossum committed:
Substitute array values -- 22 errors left

gvanrossum committed:
Skip assignment to sized arrays (these aren't used anyway)

gvanrossum committed:
If an oeffect has the same name as an ieffect, give it the same type

This fixes one error (in PUSH_EXC_INFO).
It also removes redundant dynamic pops from error returns.

gvanrossum committed:
Special-case `callable = method` -- 9 errors left!

gvanrossum committed:
Array output effects are PyObject ** -- 5 errors left

gvanrossum committed:
Don't detag(a[i]) in LHS -- still 5 errors left

gvanrossum committed:
FORMAT_VALUE

gvanrossum committed:
CALL_FUNCTION_EX

gvanrossum committed:
CALL -- 0 errors left

gvanrossum committed:
Fix bug accidentally introduced in CALL_BOUND_METHOD_EXACT_ARGS

gvanrossum committed:
WIP, 2 warnings left

gvanrossum committed:
Killed the last warnings

gvanrossum committed:
Fix colossal mistake in plain_obj_array()

gvanrossum committed:
Revert "Dumb experiment"

It somehow caused a crash in the int conversion
which I don't want to debug right now.

This reverts commit e50b6a7fba757790956add4ff1734861bcadd33b.

gvanrossum committed:
Make localsplus a _tagged_ptr[]

gvanrossum committed:
Fix lldebug code for _tagged_ptr

gvanrossum committed:
Check for STEAL() in generator

gvanrossum committed:
More STEAL() calls, mostly

gvanrossum committed:
More tweaks

lpereira commented on Include/internal/pycore_frame.h:
```diff
@@ -23,12 +23,76 @@ struct _frame {
+}
+
+static inline void incref_if_tagged(_tagged_ptr tp) {
+    if (is_tagged(tp)) {
+        Py_INCREF(detag(tp));
+    }
```
Have you seen the mostly branch-free alternative to this I [proposed here](https://github.com/faster-cpython/ideas/issues/443#issuecomment-1209736091)? (I'm not sure if it's better, though, as it changes one branch to guard a memory write to an actual memory write, but it's something to compare and see.)

The primitive for increffing is branch-free and the one to decref does a pointer comparison rather than going through a struct.

gvanrossum replied:
The line
```
untagged_ob->ob_refcnt += ((uintptr_t)ob) & 1ull;
```
in your proposal would end up always doing a memory read and write, which is exactly what we're trying to avoid here.

lpereira replied:
Yes, I mentioned this in the comment.  The unref doesn't do a memory read, though, as it only compares pointers; it could be used with an incref that performs a branch.

gvanrossum replied:
I trust branch prediction more than memory latency. :-)

gvanrossum committed:
Add STEAL calls to STORE_DEREF, MAKE_FUNCTION

gvanrossum committed:
TEMPORARY: Enable tagging, disable specializations

gvanrossum committed:
Changes to make it compile on Windows

gvanrossum committed:
Clean up LOAD_CONST tagging

A remaining crash: ./python -m test test_asyncio

gvanrossum committed:
Fix BUILD_SET

gvanrossum committed:
Enable specializing LOAD_GLOBAL

gvanrossum committed:
Tag all constants except None

gvanrossum committed:
Enable specializing SEND

gvanrossum committed:
Enable specializing BINARY_SUBSCR

gvanrossum committed:
Enable specializing STORE_SUBSCR

Fix PyList_SET_ITEM() call in STORE_SUBSCR_LIST_INT

gvanrossum committed:
Enable specializing STORE_ATTR

Add STEAL() calls to specializations.
A pattern is emerging: anything without DECREF_INPUTS() is suspect.

gvanrossum committed:
Enable specializing LOAD_ATTR

Add some STEAL() calls.

gvanrossum committed:
Enable most quickening, and COMPARE_AND_BRANCH

Some super-instructions are broken, those are in #if 0 / #endif for now.

gvanrossum committed:
Enable specializing BINARY_OP

gvanrossum committed:
Enable specializing FOR_ITER

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into tagged-ptrs

gvanrossum committed:
Fix FORMAT_VALUE (0)

gvanrossum committed:
Fix the last super instructions

gvanrossum committed:
Fix RAISE_VARARGS

gvanrossum committed:
Tag all constants, even None

gvanrossum committed:
Maybe fix RERAISE (test_barrier.py still crashes)
